api/base/serializers.py

```diff
@@ -144,7 +188,87 @@ def to_representation(self, value):
+                                                    lookup_url_kwarg='node_id', read_only=True, link_type='related')
+
+    """
+    def __init__(self, view_name=None, **kwargs):
+        self.meta = kwargs.pop('meta', {})
+        self.link_type = kwargs.pop('link_type', 'url')
+        super(JSONAPIHyperlinkedRelatedField, self).__init__(view_name=view_name, **kwargs)
+
+    def to_representation(self, value):
+        """
+        Returns nested dictionary in format {'links': {'self.link_type': ... }
+
+        If no meta information, self.link_type is equal to a string containing link's URL.  Otherwise,
+        the link is represented as a links object with 'href' and 'meta' members.
+        """
+        url = super(JSONAPIHyperlinkedRelatedField, self).to_representation(value)
+        meta = _rapply(self.meta, _url_val, obj=value, serializer=self.parent)
+        return self.format_json_response(link_type=self.link_type, url=url, meta=meta)
+
+
+class JSONAPIHyperlinkedGuidRelatedField(ser.Field, HyperlinkedFieldMixin):
+    """
+    Field that returns a nested dict with the url (constructed based
+    on the object's type), optional meta information, and link_type.
+
+    Example:
+
+        target = JSONAPIHyperlinkedGuidRelatedField(link_type='related', meta={'type': 'get_target_type'})
+
+    """
+    def __init__(self, **kwargs):
+        self.meta = kwargs.pop('meta', {})
+        self.link_type = kwargs.pop('link_type', 'url')
+        super(JSONAPIHyperlinkedGuidRelatedField, self).__init__(read_only=True, **kwargs)
+
+    def get_guid_views(self, guid):
+        # get target type from guid
+        guid_object = Guid.load(guid)
+        if not guid_object:
+            raise NotFound
+        guid_referent = guid_object.referent
+
+        if isinstance(guid_referent, Node):
+            self.link_type = 'related'
+            guid_views = GUID_VIEWS['node']
+
+        elif isinstance(guid_referent, Comment):
+            self.link_type = 'self'
+            guid_views = GUID_VIEWS['comment']
+
+        else:
+            raise InvalidModelValueError('Invalid comment target.')
+
+        return guid_views
+
+    def to_representation(self, value):
+        """
+        Returns nested dictionary in format {'links': {'self.link_type': ... }
+
+        If no meta information, self.link_type is equal to a string containing link's URL.  Otherwise,
+        the link is represented as a links object with 'href' and 'meta' members.
+        """
+        guid_views = self.get_guid_views(value._id)
+        view_name = guid_views['view_name']
+        lookup_url_kwarg = guid_views['lookup_url_kwarg']
+
+        url = Link(view_name, kwargs={lookup_url_kwarg: '<_id>'}).resolve_url(value)
+        meta = _rapply(self.meta, _url_val, obj=value, serializer=self.parent)
+
+        return self.format_json_response(link_type=self.link_type, url=url, meta=meta)
```

Comment:
Entirely possible that I'm missing something but wouldn't this work just as well?

```python
    def to_representation(self, value):
        """
        Returns nested dictionary in format {'links': {'self.link_type': ... }

        If no meta information, self.link_type is equal to a string containing link's URL.  Otherwise,
        the link is represented as a links object with 'href' and 'meta' members.
        """

        meta = _rapply(self.meta, _url_val, obj=value, serializer=self.parent)

        return self.format_json_response(link_type=self.link_type, url=value.get_absolute_url(), meta=meta)
```
And let you remove the maintenance of keeping the view/kwarg mapping up at the top and you could kill get_guid_views?

The thing that brought it to my attention was that we just made registrations and collections 404 if they're requested through the nodes endpoint. They have to be requested through /v2/registrations/\<nid\>/ and /v2/collections/\<cid\>/.